### PR TITLE
EON-1826 - python test to test reward distribution corner cases after…

### PR DESCRIPTION
**Pull request by i-skrypnyk**
1. Setup forgers:
	- forger 1 - reward 1 - delegator 1 - share 500
	- forger 2 - reward 1 - delegator 2 - share 500
	- forger 3 - reward 3 - delegator 3 - share 500
	- forger 4 - reward 4 - delegator 3 - share 500
	- forger 5 - reward 5 - null 		- null 		- not upgraded, then upgraded to reward 0
	- forger 6 - reward 6 - delegator 6 - share 1000
	- forger 7 - reward 7 == delegator 7 - share 500

2. Test flow
	- advance to epoch 67
	- send funds to new address at node 1, advance to 68
	- send some zen to 6 forgers, forge required blocks to create 6 new forger stakes
	- reach fork 1.3, execute upgrade
	- test block forging, reach epoch 79
	- send 100 zen to mc_reward_pool
	- distribute fees, ignore the results
	- send 1000 zen to mc_reward_pool
	- forger stake is activated
	- forgers are upgraded
	- distribute fees
		- forger rewards:
			- forger 1 <1.4 - 2/16 of mc_reward_pool + 2/16 block_fees + 2 zennies remainder
			- forger 1 >1.4 - 3/16 of mc_reward_pool + 3/16 block_fees + tips + 2 zennies remainder
			- forger 2 		- 3/16 of mc_reward_pool + 3/16 block_fees + 2 zennies remainder
			- forger 3 		- 2/16 of mc_reward_pool + 2/16 block_fees
			- forger 4 		- 2/16 of mc_reward_pool + 2/16 block_fees
			- forger 5 		- 2/16 of mc_reward_pool + 2/16 block_fees
			- forger 6 		- 1/16 of mc_reward_pool + 1/16 block_fees
			- forger 7 		- 1/16 of mc_reward_pool + 1/16 block_fees

		- reward 1 gets
			- 100% from forger 1 before fork
			- 50% from forger 1 after fork
			- 50% from forger 2
		- reward 3 gets
			- 50% from forger 3
		- reward 4 gets
			- 50% from forger 4
		- reward 5 gets
			- 100% from forger 5
		- reward 6 gets 0
		- reward 7 gets
			- 100% from forger 7(as it is == to delegator 7)

		- delegator 1 gets
			- 50% from forger 1 after fork
		- delegator 2 gets
			- 50% from forger 2
		- delegator 3 gets
			- 50% from forger 3
			- 50% from forger 4
		- delegator 6 gets
			- 100% from forger 6
		- delegator 7 gets
			- 100% from forger 7(as it is == to reward 7)

### CODE REVIEW

File: qa/sc_evm_forger_and_delegator_rewards.py [hunk 1]
```diff
@@ -0,0 +1,525 @@
...
+
+    def activate_stake_v2(self, evm_address_sc_node_1, sc_node_1):
+        forger_v2_native_contract = SmartContract("ForgerStakesV2")
+        method = 'activate()'
+        tx_hash = contract_function_call(sc_node_1, forger_v2_native_contract, FORGER_STAKE_V2_SMART_CONTRACT_ADDRESS,
+                                         evm_address_sc_node_1, method)
+        generate_next_block(sc_node_1, "first node")
+        _, forgersPoolFee, forgerTip = computeForgedTxFee(sc_node_1, tx_hash)
+        self.sc_sync_all()
+        generate_next_block(sc_node_1, "second node")
```
Comment (paolocappelletti) [inline qa/sc_evm_forger_and_delegator_rewards.py:378]: typo: second node => first node

File: qa/sc_evm_forger_and_delegator_rewards.py [hunk 2]
```diff
@@ -0,0 +1,525 @@
...
+        self.sc_sync_all()
+
+        # Check that the fee distribution took place, exact fees before 1.4 are not tested here
+        api_fee_payments = http_block_getFeePayments(sc_node_1, sc_last_we_block_id)['feePayments']
+        # 7 because the genesis block is counted towards node 1 forger address, not reward address
+        assert_equal(7, len(api_fee_payments))
+
+        # trigger cert submission
+        # Generate 2 SC blocks on SC node and start them automatic cert creation.
+        generate_next_block(sc_node_1, "first node")  # 1 SC block to reach the end of WE
```
Comment (paolocappelletti) [inline qa/sc_evm_forger_and_delegator_rewards.py:214]: isn't the comment misleading? we should have reached already the end of WE with the previous block, at line 204

File: qa/sc_evm_forger_and_delegator_rewards.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+        generate_next_block(sc_node_1, "first", force_switch_to_next_epoch=True)  # 80
+        ft_pool_amount = 100
+        forward_transfer_to_sidechain(self.sc_nodes_bootstrap_info.sidechain_id,
+                                      mc_node,
+                                      format_eoa(FORGER_POOL_RECIPIENT_ADDRESS),
+                                      ft_pool_amount,
+                                      mc_return_address=mc_node.getnewaddress(),
+                                      generate_block=False)
+        # Activate forger stake v2
+        forger_pool_fee_1, node_1_tip_1 = self.activate_stake_v2(evm_address_sc_node_1, sc_node_1)
```
Comment (paolocappelletti) [inline qa/sc_evm_forger_and_delegator_rewards.py:239]: for bettere readability let's add a comment saying this method internally generates 2 more blocks

**Review by paolocappelletti — CHANGES_REQUESTED** [review]
looks good, just some minor changes

File: qa/sc_evm_forger_and_delegator_rewards.py (hunk 2 above)
Comment (i-skrypnyk) [reply qa/sc_evm_forger_and_delegator_rewards.py:214]: I don't really know how this works, copied this part from other test. Will remove the comment completely

**Review by saratnt — APPROVED** [review]
LGTM! Just some typos.
For the problem generated by the ActiveSlotCoefficient, maybe it can be created another command line argument in the EvmApp for disabling it, like it was done for --allforks. It takes some work but it would be very useful for the python tests. @paolocappelletti maybe it can be created a task in Jira for that?

File: qa/sc_evm_forger_and_delegator_rewards.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+        sc_node_5 = self.sc_nodes[4]
+        sc_node_6 = self.sc_nodes[5]
+        sc_node_7 = self.sc_nodes[6]
+        connect_sc_nodes(sc_node_1, 1)
+        self.sc_sync_all()
+        self.advance_to_epoch(67)
+        self.sync_all()
+        self.sc_sync_all()
+
+        # mc_node.generate(self.withdrawalEpochLength)
```
[the remark this answers is not in the document]
Comment (i-skrypnyk) [reply qa/sc_evm_forger_and_delegator_rewards.py:125]: Thanks, I was sure I removed that, that was me trying to make blocks forge more consistently, but it had no effect

**Review by paolocappelletti — APPROVED** [review]
